controller/tap/apiserver.go

```diff
@@ -0,0 +1,342 @@
+	server.router.GET("/apis/"+gvk.GroupVersion().String(), server.handleAPIResourceList)
+
+	for _, res := range resources {
+		route := ""
+		if !res.namespaced {
+			route = fmt.Sprintf("/apis/%s/watch/%s/:namespace/tap", gvk.GroupVersion().String(), res.name)
+		} else {
+			route = fmt.Sprintf("/apis/%s/watch/namespaces/:namespace/%s/:name/tap", gvk.GroupVersion().String(), res.name)
+		}
+		server.router.POST(route, server.handleTap)
+	}
+
+	clientCertPool := x509.NewCertPool()
+	clientCertPool.AppendCertsFromPEM([]byte(clientCAPem))
+
+	wrappedServer := prometheus.WithTelemetry(server)
+
+	s := &http.Server{
+		Addr:    addr,
+		Handler: wrappedServer,
+		TLSConfig: &tls.Config{
+			Certificates: []tls.Certificate{cert},
+			ClientAuth:   tls.VerifyClientCertIfGiven,
+			ClientCAs:    clientCertPool,
+		},
+	}
+
+	lis, err := net.Listen("tcp", addr)
+	if err != nil {
+		server.log.Fatalf("net.Listen failed with: %s", err)
+	}
+
+	return s, lis, nil
+}
+
+// ServeHTTP handles all routes for the APIServer.
+func (a *apiServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	a.log.Debugf("ServeHTTP(): %+v", req)
+
+	// if `requestheader-allowed-names` was empty, allow any CN
+	if len(a.allowedNames) > 0 {
+		validCN := ""
+		clientNames := []string{}
+		for _, cn := range a.allowedNames {
+			for _, clientCert := range req.TLS.PeerCertificates {
+				clientNames = append(clientNames, clientCert.Subject.CommonName)
+				if cn == clientCert.Subject.CommonName {
+					validCN = clientCert.Subject.CommonName
+					break
+				}
+			}
+			if validCN != "" {
+				break
+			}
+		}
+		if validCN == "" {
+			err := fmt.Errorf("no valid CN found. allowed names: %s, client names: %s", a.allowedNames, clientNames)
+			a.log.Debug(err)
+			renderJSONError(w, err, http.StatusBadRequest)
+			return
+		}
+	}
+
+	a.router.ServeHTTP(w, req)
+}
+
+// /apis/tap.linkerd.io/v1alpha1
+// this is required for `kubectl api-resources` to work
+func (a *apiServer) handleAPIResourceList(w http.ResponseWriter, req *http.Request, p httprouter.Params) {
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(a.apiResList)
+}
+
+// /apis/tap.linkerd.io/v1alpha1/watch/namespaces/:namespace/tap
+// /apis/tap.linkerd.io/v1alpha1/watch/namespaces/:namespace/:resource/:name/tap
+func (a *apiServer) handleTap(w http.ResponseWriter, req *http.Request, p httprouter.Params) {
+	namespace := p.ByName("namespace")
+	name := p.ByName("name")
+	resource := ""
+
+	path := strings.Split(req.URL.Path, "/")
+	if len(path) == 8 {
+		resource = path[5]
+	} else if len(path) == 10 {
+		resource = path[7]
+	} else {
+		err := fmt.Errorf("invalid path: %s", req.URL.Path)
+		a.log.Error(err)
+		renderJSONError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	a.log.Debugf("SubjectAccessReview: namespace: %s, resource: %s, name: %s, user: %s, group: %s",
+		namespace, resource, name, req.Header.Get(a.usernameHeader), req.Header[a.groupHeader],
+	)
+
+	err := pkgK8s.ResourceAuthzForUser(
```

Comment:
Is this subject access review necessary? are there requests which could get past the aggregator but fail here?

Reply:
I haven't been able to craft a request that gets by the aggregator but fails this SAR, but I'm not convinced it's not possible. This is the recommended pattern from Kubernetes:
https://kubernetes.io/docs/tasks/access-kubernetes-api/configure-aggregation-layer/#extension-apiserver-authorizes-the-request

Reply:
I'm pretty sure this is the exact same SAR check that the aggregator does.  So doing it here as well feels redundant.  Removing it also allows us to remove the auth-delegator role binding.